zipline/pipeline/loaders/blaze/core.py

```diff
@@ -925,17 +1002,39 @@ def collect_expr(e):
             This can return more data than needed. The in memory reindex will
             handle this.
             """
-            df = odo(where(e), pd.DataFrame, **odo_kwargs)
-            df.sort(TS_FIELD_NAME, inplace=True)  # sort for the groupby later
-            return df
-
-        materialized_expr = collect_expr(expr)
+            predicate = e[TS_FIELD_NAME] <= upper_dt
+            if lower is not None:
+                predicate &= e[TS_FIELD_NAME] >= lower
+
+            return odo(e[predicate][colnames], pd.DataFrame, **odo_kwargs)
+
+        if checkpoints is not None:
+            ts = checkpoints[TS_FIELD_NAME]
+            checkpoints_ts = odo(ts[ts <= lower_dt].max(), pd.Timestamp)
+            if pd.isnull(checkpoints_ts):
+                materialized_checkpoints = pd.DataFrame(columns=colnames)
+                lower = lower_dt
```

Comment:
Ah, I think this case needs to be `lower = None` because this means we have no checkpoints so we should just query everything.